### PR TITLE
Fix #389: Home Fragment Carousel [DO NOT MERGE]

### LINKED ISSUE #389: Replace promoted story card on home page with carousel
The home page should use a carousel to allow horizontal scrolling across all recent/promoted stories rather than only showing the top item and requiring users to click into the 'view all' page to see others.

### PR DESCRIPTION
**Pull request by rt4914**
## Explanation

This is just a rough implementation of Home Fragment Carousel. Do not check the code quality just focus on the working on HomeFragment.

home_fragment

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt [hunk 1]
```diff
@@ -146,4 +147,38 @@ class HomeFragmentPresenter @Inject constructor(
...
     routeToTopicListener.routeToTopic(topicSummary.topicId)
   }
+
+  private val ongoingStoryListSummaryResultLiveData: LiveData<AsyncResult<OngoingStoryList>> by lazy {
+    topicListController.getOngoingStoryList()
+  }
+
+  private fun subscribeToPromotedStoryList() {
+    getAssumedSuccessfulOngoingStoryList().observe(fragment, Observer<OngoingStoryList> {
+      if (it.recentStoryCount > 0) {
```
Comment (jamesxu0) [inline app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:157]: Is this "if" check necessary? If recentStoryCount is empty, the for loop shouldn't do anything right? Same for below.

File: app/src/main/java/org/oppia/app/home/topiclist/PromotedStoryListAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package org.oppia.app.home.topiclist
+
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import org.oppia.app.databinding.PromotedStoryCardBinding
+
+/** Adapter to inflate different items/views inside [RecyclerView]. The itemList consists of various ViewModels. */
+class PromotedStoryListAdapter(
```
Comment (jamesxu0) [inline app/src/main/java/org/oppia/app/home/topiclist/PromotedStoryListAdapter.kt:9]: Is it necessary to create a separate adapter? I feel like using BindableAdapter would suffice.

**Comment by seanlip** [comment]
Hi @rt4914, thanks for this!

Just one design note: I checked with @mschanteltc and the intended behaviour is for the tile to "anchor" once the swiping is stopped -- i.e. it should not be possible to have the tile boundary sit midway on the screen when the finger is removed. Instead, the tile should snap/spring to the left/right so that the leftmost tile on the screen is always the same location (when there's no finger on the screen). Does this make sense?

File: app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt (hunk 1 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:157]: Oops. Sorry, I didn't read your explanation correctly.

File: app/src/main/java/org/oppia/app/home/topiclist/PromotedStoryListAdapter.kt (hunk 2 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/home/topiclist/PromotedStoryListAdapter.kt:9]: Oops. Sorry, I didn't read your explanation correctly.

File: app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt (hunk 1 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:157]: No issues, I will keep this comment for now, and will solve this.

File: app/src/main/java/org/oppia/app/home/topiclist/PromotedStoryListAdapter.kt (hunk 2 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/home/topiclist/PromotedStoryListAdapter.kt:9]: No issues, I will keep this comment for now, and will solve this.

**Comment by rt4914** [comment]
> Hi @rt4914, thanks for this!
> 
> Just one design note: I checked with @mschanteltc and the intended behaviour is for the tile to "anchor" once the swiping is stopped -- i.e. it should not be possible to have the tile boundary sit midway on the screen when the finger is removed. Instead, the tile should snap/spring to the left/right so that the leftmost tile on the screen is always the same location (when there's no finger on the screen). Does this make sense?

Yes that makes sense. I will fix that once this PR gets a heads up from @BenHenning too.

**Comment by rt4914** [comment]
@BenHenning In this PR, please check the code approach, it is somewhat similar to the approach that we following in TopicTrainTab, using NestedRecyclerView. Also, I have tested this app on lower APIs too and it works fine.

**Comment by rt4914** [comment]
@BenHenning after reading more about this on internet, I found that using NestedScrollView and then keeping `recyclerView.isNestedScrollingEnabled = false` works for our case but defeats the entire purpose of recyclerview because now recyclerview will **not** recycle the views.

**Comment by BenHenning** [comment]
Hmm, that is definitely not ideal. I wouldn't mind if were losing recycling for the carousel, but for the entire home fragment that won't work. We could potentially be showing dozens of topics here, so we definitely want recycling functionality.

**Comment by rt4914** [comment]
> Hmm, that is definitely not ideal. I wouldn't mind if were losing recycling for the carousel, but for the entire home fragment that won't work. We could potentially be showing dozens of topics here, so we definitely want recycling functionality.

Exactly, so I believe this won't be an ideal solution too. We will have to find out something else.

**Comment by BenHenning** [comment]
Does it work to model the carousel as its own recycler view within the outer recycler view? Given that we have two directions of scrolling, that might work without the additional nested scrolling being necessary. It makes the recycling of the carousel really expensive, but that probably will work better than the performance loss of disabling recycling on the whole view. We *might* have some scroll state saving issues, though, and may need to work around that (e.g. by caching the position of the carousel when it gets recycled).

**Comment by rt4914** [comment]
> Does it work to model the carousel as its own recycler view within the outer recycler view? Given that we have two directions of scrolling, that might work without the additional nested scrolling being necessary. It makes the recycling of the carousel really expensive, but that probably will work better than the performance loss of disabling recycling on the whole view. We _might_ have some scroll state saving issues, though, and may need to work around that (e.g. by caching the position of the carousel when it gets recycled).

Makes sense, lets try this out and see what happens. This will be somewhat similar to TopicPlayTab nested recyclerview.

File: app/src/main/java/org/oppia/app/home/topiclist/PromotedStoryViewModel.kt [hunk 3]
```diff
@@ -11,12 +11,12 @@ import org.oppia.app.home.RouteToTopicPlayStoryListener
 import org.oppia.app.home.continueplaying.ContinuePlayingActivity
```
Comment (veena14cs) [inline app/src/main/java/org/oppia/app/home/topiclist/PromotedStoryViewModel.kt:11]: Remove unused import.

File: app/src/main/java/org/oppia/app/home/topiclist/TopicSummaryViewModel.kt [hunk 4]
```diff
@@ -3,6 +3,7 @@ package org.oppia.app.home.topiclist
 import android.graphics.Color
 import android.view.View
 import androidx.annotation.ColorInt
+import androidx.lifecycle.ViewModel
 import org.oppia.app.home.HomeItemViewModel
```
Comment (veena14cs) [inline app/src/main/java/org/oppia/app/home/topiclist/TopicSummaryViewModel.kt:7]: Remove unused imports.

**Comment by nikitamarysolomanpvt** [comment]
Please assign back to me once review is required...

**Comment by rt4914** [comment]
> Does it work to model the carousel as its own recycler view within the outer recycler view? Given that we have two directions of scrolling, that might work without the additional nested scrolling being necessary. It makes the recycling of the carousel really expensive, but that probably will work better than the performance loss of disabling recycling on the whole view. We _might_ have some scroll state saving issues, though, and may need to work around that (e.g. by caching the position of the carousel when it gets recycled).

Solved in #477

**Comment by rt4914** [comment]
> Hi @rt4914, thanks for this!
> 
> Just one design note: I checked with @mschanteltc and the intended behaviour is for the tile to "anchor" once the swiping is stopped -- i.e. it should not be possible to have the tile boundary sit midway on the screen when the finger is removed. Instead, the tile should snap/spring to the left/right so that the leftmost tile on the screen is always the same location (when there's no finger on the screen). Does this make sense?

@seanlip Solved in #477

**Comment by rt4914** [comment]
Recreated this with alternative approach in #477

**Pull request closed without merging**